crates/language/src/buffer.rs

```diff
@@ -3058,6 +3089,241 @@ impl BufferSnapshot {
+                            if trailing_row > self.max_point().row
+                                || trailing_row > end_row + TRAILING_ROW_SEARCH_LIMIT
+                            {
+                                break;
+                            }
+                            let (new_line_indent, empty) = self.line_indent_for_row(trailing_row);
+                            (trailing_row, new_line_indent, empty)
+                        };
+
+                    if empty {
+                        continue;
+                    }
+                    last_row = target_row.min(end_row);
+                    line_indent = new_line_indent;
+                    found_indent = true;
+                    break;
+                }
+            } else {
+                found_indent = true
+            }
+
+            let depth = if found_indent {
+                line_indent / indent_size + ((line_indent % indent_size) > 0) as u32
+            } else {
+                current_depth
+            };
+
+            if depth < current_depth {
+                for _ in 0..(current_depth - depth) {
+                    let (mut range, mut indent) = indent_stack.pop().unwrap();
+                    if last_row != first_row {
+                        // In this case, we landed on an empty row, had to seek forward,
+                        // and discovered that the indent we where on is ending.
+                        // This means that the last display row must
+                        // be on line that ends this indent range, so we
+                        // should display the range up to the first non-empty line
+                        indent.end_row = first_row.saturating_sub(1);
+                        range.end = Point::new(indent.end_row, 0).to_offset(self);
+                    }
+
+                    result_vec.push((range, indent))
+                }
+            } else if depth > current_depth {
+                for next_depth in current_depth..depth {
+                    let range = Point::new(first_row, 0).to_offset(self)
+                        ..Point::new(last_row, 0).to_offset(self);
+
+                    indent_stack.push((
+                        range,
+                        IndentGuide {
+                            buffer_id: self.remote_id(),
+                            start_row: first_row,
+                            end_row: last_row,
+                            depth: next_depth,
+                            indent_size,
+                        },
+                    ));
+                }
+            }
+
+            for (range, indent) in indent_stack.iter_mut() {
+                indent.end_row = last_row;
+                range.end = Point::new(indent.end_row, 0).to_offset(self);
+            }
+        }
+
+        result_vec.extend(indent_stack);
+
+        result_vec
+    }
+
+    pub async fn enclosing_indent(&self, mut buffer_row: u32) -> Option<(Range<u32>, u32)> {
```

Comment:
> u32, u32, 32

Passing by in curiosity, and wanted to note that using `MultiBufferRow` and `DisplayRow` might fix panics and reduce bugs.
Would be nice to remove all u32's possible around, but not insisting.